libmamba/src/core/logging.cpp

```diff
@@ -0,0 +1,263 @@
+
+#include <iostream>
+#include <mutex>
+#include <shared_mutex>
+#include <utility>
+#include <vector>
+
+#include <fmt/core.h>  // TODO: replace by `<format>` once available on all ci compilers
+
+#include <mamba/core/context.hpp>
+#include <mamba/core/error_handling.hpp>
+#include <mamba/core/execution.hpp>
+#include <mamba/core/invoke.hpp>
+#include <mamba/core/logging.hpp>
+#include <mamba/core/output.hpp>  // TODO: remove
+#include <mamba/core/util.hpp>
+#include <mamba/util/synchronized_value.hpp>
+
+namespace mamba::logging
+{
+
+    namespace details
+    {
+        // see singletons.cpp
+#if defined(__APPLE__)
+        using params_mutex = std::mutex;
+#else
+        using params_mutex = std::shared_mutex;
+#endif
+        extern util::synchronized_value<LoggingParams, params_mutex> logging_params;
+        extern AnyLogHandler current_log_handler;
+    }
+
+    namespace
+    {
+        // TODO: consider generalize and move in synchronized_value.hpp
+        template <std::default_initializable T, typename U, typename... OtherArgs>
+            requires std::assignable_from<T&, U>
+        auto synchronize_with_value(
+            util::synchronized_value<T, OtherArgs...>& sv,
+            const std::optional<U>& new_value
+        )
+        {
+            auto synched_value = sv.synchronize();
+            if (new_value)
+            {
+                *synched_value = *new_value;
+            }
+            return synched_value;
+        }
+    }
+
+    AnyLogHandler::~AnyLogHandler()
+    {
+        // We handle the case where we don't exit normally (calling `exit(0);` for example)
+        // but we still need to properly end the logging system.
+        // If `this` is the currently registered log-handler, stop it properly before
+        // calling the implementation's destruction.
+        if (has_value() and this == &details::current_log_handler)
+        {
+            static std::once_flag flag;
+
+            // clang-format off
+            std::call_once(flag, [&]{
+                // We dont want to propagate the error if any, just log it and continue;
+                // we dont need the resulting value if any.
+                [[maybe_unused]] auto result = safe_invoke([this] { this->stop_log_handling(stop_reason::program_exit); })
+                    .map_error([](const mamba_error& error){
+                            // Here with report the error in the standard output to avoid any logging
+                            // implementation.
+                            const auto message = fmt::format(
+                                "mamba::logging termination failure: call to `stop_log_handling()` ended with an error (caught, logged, skipped): {}",
+                                error.what()
+                            );
+                            std::cerr << message << std::endl;
+                            std::cout << message << std::endl;
```

Comment:
Both?

Reply:
I think std::cerr is the stream made for reporting error here. Which workflow could make this invisible if we report on std::cerr only?

Reply:
Not certain, I remember issues with only using `std::cerr` but cant find what it was about.
I'll use `std::cerr` only, we can easily change that later if necessary.